### PR TITLE
improvements on initial sysroot and libdir finding logics

### PR DESCRIPTION
**Pull request by onur-ozkan**
Stabilized initial sysroot and libdir path resolution logic to work without dry-run conditions and utilized initial sysroot more broadly.

### CODE REVIEW

File: src/bootstrap/src/lib.rs [hunk 1]
```diff
@@ -315,46 +315,19 @@ impl Build {
...
-                rustc = config.initial_rustc,
-            );
-        };
+        let initial_libdir = initial_target_dir
+            .ancestors()
+            .nth(2)
+            .unwrap()
+            .strip_prefix(&config.initial_sysroot)
+            .unwrap()
+            .to_path_buf();
```
Comment (onur-ozkan) [inline src/bootstrap/src/lib.rs:330]: I have never experienced this failing before and it should be even more stable now.

cc @Zalathar

Comment (Zalathar) [reply src/bootstrap/src/lib.rs:330]: I still see this fail occasionally (on macOS only), but I have already collected plenty of examples of the failure so I don’t really need this code any more.

If you replace the unwrap with something like `.expect("couldn’t determine initial libdir")`, I’m fine with that.

Comment (Zalathar) [reply src/bootstrap/src/lib.rs:330]: For reference, the failure looks like this: https://github.com/rust-lang/rust/pull/130138#issuecomment-2402056712

Two different invocations of the same compiler binary produce different sysroot paths. It probably has something to do with hardlink bookkeeping on macOS specifically.

Comment (Zalathar) [reply src/bootstrap/src/lib.rs:330]: Oh I see above that we aren’t invoking rustc multiple times anymore, so yeah from a bootstrap perspective this should avoid the problem I was seeing.

Comment (onur-ozkan) [reply src/bootstrap/src/lib.rs]: > If you replace the unwrap with something like .expect("couldn’t determine initial libdir"), I’m fine with that.

Done that in https://github.com/rust-lang/rust/compare/16f30f2f2d92ee450e4ec715ee09e17dc2c1f5b8..27323aac9fdb905e0665fc9c43147a895ef51da6.

**Comment by onur-ozkan** [comment]
r? bootstrap

**Review by jieyouxu — APPROVED** [review]
Thanks, this looks like a nice cleanup.

**Pull request merged**